thread-manager/src/tokio_runtime.rs

```diff
@@ -0,0 +1,178 @@
+
+        let base_name = name.clone();
+        let mut builder = match num_workers {
+            1 => tokio::runtime::Builder::new_current_thread(),
+            _ => {
+                let mut builder = tokio::runtime::Builder::new_multi_thread();
+                builder.worker_threads(num_workers);
+                builder
+            }
+        };
+        let atomic_id: AtomicUsize = AtomicUsize::new(0);
+
+        let counters = Arc::new(ThreadCounters {
+            namespace: format!("thread-manager-tokio-{}", &base_name).leak(), // no workaround, metrics crate will only consume 'static str
+            parked_threads_cnt: AtomicI64::new(0),
+            active_threads_cnt: AtomicI64::new(
+                (num_workers.wrapping_add(cfg.max_blocking_threads)) as i64,
+            ),
+        });
+        let counters_clone1 = counters.clone();
+        let counters_clone2 = counters.clone();
+        builder
+            .event_interval(cfg.event_interval)
+            .thread_name_fn(move || {
+                let id = atomic_id.fetch_add(1, Ordering::Relaxed);
+                format!("{}-{}", base_name, id)
+            })
+            .on_thread_park(move || {
+                counters_clone1.on_park();
+            })
+            .on_thread_unpark(move || {
+                counters_clone2.on_unpark();
+            })
+            .thread_stack_size(cfg.stack_size_bytes)
+            .enable_all()
+            .max_blocking_threads(cfg.max_blocking_threads);
+
+        //keep borrow checker happy and move these things into the closure
+        let c = cfg.clone();
+        let chosen_cores_mask = Mutex::new(chosen_cores_mask);
+        builder.on_thread_start(move || {
+            let cur_thread = std::thread::current();
+            let _tid = cur_thread
+                .get_native_id()
+                .expect("Can not get thread id for newly created thread");
+            // todo - tracing
+            //let tname = cur_thread.name().unwrap();
+            //println!("thread {tname} id {tid} started");
+            apply_policy(
+                &c.core_allocation,
+                parse_policy(&c.policy),
+                c.priority,
+                &chosen_cores_mask,
+            );
+        });
+        Ok(TokioRuntime {
+            tokio: builder.build()?,
+            config: cfg.clone(),
+            counters,
+        })
+    }
+
+    /// Makes test runtime with 2 threads, only for unittests
+    pub fn new_for_tests() -> Self {
+        let cfg = TokioConfig {
+            worker_threads: 2,
+            ..Default::default()
+        };
+        TokioRuntime::new("solNetTest".to_owned(), cfg.clone()).unwrap()
+    }
+}
+
+///Internal counters to keep track of worker pool utilization
+#[derive(Debug)]
+pub struct ThreadCounters {
+    pub namespace: &'static str,
+    pub parked_threads_cnt: AtomicI64,
+    pub active_threads_cnt: AtomicI64,
```

Comment:
bit of a nit, but this seems like an odd way to do this? there's subtle synchronization issues with the 2 atomics. we should just have a `total_thread_count: u64` which is unchanging, and then `active_threads_cnt: AtomicU64` (these aren't ever negative so don't think they need to be i64?)

`parked_threads_cnt` can be derived from `total_thread_count - active_thread_count.load()`, but now theres' no weird edge case where the reporting loads:
```rust
        let parked = counters.parked_threads_cnt.load(Ordering::Relaxed);
        let active = counters.active_threads_cnt.load(Ordering::Relaxed);
```
load just as `on_park` or `on_unpark` is being run and the numbers don't add up correctly.

Reply:
yes, that is a good idea!